rust/src/io/commit/dynamodb.rs

```diff
@@ -0,0 +1,492 @@
+            }),
+        }
+    }
+
+    /// Get the latest version of a dataset at the base_uri
+    async fn get_latest_version(&self, base_uri: &str) -> Result<Option<u64>> {
+        let query_result = self
+            .query()
+            .key_condition_expression(format!("{BASE_URI} = :{BASE_URI}"))
+            .expression_attribute_values(format!(":{BASE_URI}"), AttributeValue::S(base_uri.into()))
+            .scan_index_forward(false)
+            .limit(1)
+            .send()
+            .await?;
+
+        match query_result.items {
+            Some(mut items) => {
+                if items.is_empty() {
+                    return Ok(None);
+                }
+                if items.len() > 1 {
+                    // this shouldn't happen, but AWS sdk doesn't model this nicely
+                    return Err(Error::IO {
+                        message: format!(
+                            "dynamodb table: {} return unexpect number of items",
+                            self.table_name
+                        ),
+                    });
+                }
+
+                let item = items.pop().expect("length checked");
+                let version_attibute = item
+                .get(VERSION)
+                .context(
+                    IOSnafu {
+                        message: format!("dynamodb error: found entries for {base_uri} but the returned data does not contain {VERSION} column")
+                    }
+                )?;
+                match version_attibute {
+                    AttributeValue::N(version) => Ok(Some(
+                        version.parse().map_err(|e| Error::IO {message: format!("dynamodb error: could not parse the version number returned {}, error: {}", version, e)})?
+                    )),
+                    _ => Err(Error::IO {
+                        message: format!("dynamodb error: found entries for {base_uri} but the returned data is not number type")
+                    })
+                }
+            }
+            _ => Ok(None),
+        }
+    }
+
+    /// Put the manifest path for a given base_uri and version, should fail if the version already exists
+    async fn put_if_not_exists(&self, base_uri: &str, version: u64, path: String) -> Result<()> {
+        self.put()
+            .item(BASE_URI, AttributeValue::S(base_uri.into()))
+            .item(VERSION, AttributeValue::N(version.to_string()))
+            .item(PATH, AttributeValue::S(path))
+            .condition_expression(format!(
+                "attribute_not_exists({BASE_URI}) AND attribute_not_exists({VERSION})",
+            ))
+            .send()
+            .await?;
+
+        Ok(())
+    }
+}
+
+// TODO: these tests are copied from super::external_manifest::test
+// since these tests applies to all external manifest stores,
+// we should move them to a common place
+
+#[cfg(all(test, target_os = "linux", feature = "dynamodb_tests"))]
```

Comment:
This wont run on mac?

Reply:
the host for `mac-build` doesn't have docker installed

Reply:
added comments for this